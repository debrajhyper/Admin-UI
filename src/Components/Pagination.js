import React from 'react'

import { HiChevronDoubleLeft, HiChevronDoubleRight } from 'react-icons/hi';
import { IoIosArrowBack, IoIosArrowForward } from 'react-icons/io';

const Pagination = ({ users, itemPerPage, currentPage,  setCurrentPage }) => {

    //  Array of numbers containing the number of pages.
    var pages = [];

    //  Calculate the number of pages.
    for(let i=1; i<=Math.ceil(users.length/itemPerPage); i++) {
        pages.push(i);
    };

    //  set the current page to the clicked page number.
    const handleClick = e => {
        setCurrentPage(Number(e.target.id));
    };

    //  paginate to the first page.
    const handleFirstBtn = () => {
        setCurrentPage(1);
    };

    //  paginate to the previous page.
    const handlePrevBtn = () => {
        setCurrentPage(currentPage - 1);
    };

    //  page numbers to be displayed in the pagination.
    const renderPageNumbers = pages.map(number => {
        return (
            <li key={number}>
                <button 
                    id={number} 
                    className={`${currentPage === number ? "active" : null} w-8 md:w-10 h-8 md:h-10 bg-white text-indigo-600 transition-colors duration-100 rounded-full focus:shadow-outline hover:bg-indigo-100`} 
                    onClick={handleClick}
                >
                    {number}
                </button>
            </li>
        );
    });

    //  paginate to the next page.
    const handleNextBtn = () => {
        setCurrentPage(currentPage + 1);
    };

    //  paginate to the last page.
    const handleLastBtn = () => {
        setCurrentPage(Math.ceil(users.length/itemPerPage));
    };

    return (
        <nav id="pagination" aria-label="Page navigation">
            <ul className="inline-flex space-x-1 sm:space-x-2 md:space-x-4 px-auto text-center">
                <li>
                    <button 
                    id="first" 
                    className="flex items-center justify-center w-8 md:w-10 h-8 md:h-10 text-indigo-600 transition-colors duration-150 rounded-full focus:shadow-outline hover:bg-indigo-100 disabled:opacity-60 disabled:bg-transparent" 
                    onClick={handleFirstBtn} 
                    disabled={currentPage === pages[0] ? true : false} 
                    >
                    <HiChevronDoubleLeft />
                    </button>
                </li>
                <li>
                    <button 
                    id="prev" 
                    className="flex items-center justify-center w-8 md:w-10 h-8 md:h-10 text-indigo-600 transition-colors duration-150 rounded-full focus:shadow-outline hover:bg-indigo-100 disabled:opacity-60 disabled:bg-transparent" 
                    onClick={handlePrevBtn} 
                    disabled={currentPage === pages[0] ? true : false}
                    >
                    <IoIosArrowBack />
                    </button>
                </li>
                {renderPageNumbers}
                <li>
                    <button 
                    id="next" 
                    className="flex items-center justify-center w-8 md:w-10 h-8 md:h-10 text-indigo-600 transition-colors duration-150 rounded-full focus:shadow-outline hover:bg-indigo-100 disabled:opacity-60 disabled:bg-transparent" 
                    onClick={handleNextBtn} 
                    disabled={currentPage === pages[pages.length - 1] ? true : false} 
                    >
                    <IoIosArrowForward />
                    </button>
                </li>
                <li>
                    <button 
                    id="last" 
                    className="flex items-center justify-center w-8 md:w-10 h-8 md:h-10 text-indigo-600 transition-colors duration-150 rounded-full focus:shadow-outline hover:bg-indigo-100 disabled:opacity-60 disabled:bg-transparent" 
                    onClick={handleLastBtn} 
                    disabled={currentPage === pages[pages.length - 1] ? true : false}
                    >
                    <HiChevronDoubleRight />
                    </button>
                </li>
            </ul>
        </nav>
    )
}

export default Pagination
