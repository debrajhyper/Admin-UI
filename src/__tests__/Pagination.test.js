import { render, screen, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import Pagination from "../Components/Pagination";

afterEach(cleanup);

test("render without crashing", () => {
    const users = [
        { id: 1, name: "Aaron Miles", email: "aaron@mailinator.com", role: "member" },
        { id: 2, name: "Aishwarya Naik", email: "aishwarya@mailinator.com", role: "member" },
        { id: 3, name: "Arvind Kumar", email: "arvind@mailinator.com", role: "admin" },
    ];
    render(<Pagination filteredUsers={users} itemPerPage={10} currentPage={1} />);

    const PaginationElement = screen.getByTestId("pagination");
    expect(PaginationElement).toBeInTheDocument();
    expect(PaginationElement).toContainHTML("<nav");
    expect(PaginationElement).toContainHTML("<ul");
    expect(PaginationElement).toContainHTML("<li");
})

test("snapshot", () => {
    const users = [
        { id: 1, name: "Aaron Miles", email: "aaron@mailinator.com", role: "member" },
        { id: 2, name: "Aishwarya Naik", email: "aishwarya@mailinator.com", role: "member" },
        { id: 3, name: "Arvind Kumar", email: "arvind@mailinator.com", role: "admin" },
    ];

    const tree = renderer.create(<Pagination filteredUsers={users} itemPerPage={10} currentPage={1} />).toJSON();
    expect(tree).toMatchSnapshot();
})