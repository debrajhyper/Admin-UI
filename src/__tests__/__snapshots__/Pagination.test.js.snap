// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<nav
  aria-label="Page navigation"
  data-testid="pagination"
  id="pagination"
>
  <ul
    className="inline-flex space-x-1 sm:space-x-4 md:space-x-4 px-auto text-center"
  >
    <li>
      <button
        className="flex items-center justify-center w-10 h-10 text-indigo-600 transition-colors duration-150 rounded-full focus:shadow-outline hover:bg-indigo-100 dark:hover:bg-indigo-300 disabled:opacity-60 disabled:bg-transparent"
        disabled={true}
        id="first"
        name="first"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M15.707 15.707a1 1 0 01-1.414 0l-5-5a1 1 0 010-1.414l5-5a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 010 1.414zm-6 0a1 1 0 01-1.414 0l-5-5a1 1 0 010-1.414l5-5a1 1 0 011.414 1.414L5.414 10l4.293 4.293a1 1 0 010 1.414z"
            fillRule="evenodd"
          />
        </svg>
      </button>
    </li>
    <li>
      <button
        className="flex items-center justify-center w-10 h-10 text-indigo-600 transition-colors duration-150 rounded-full focus:shadow-outline hover:bg-indigo-100 dark:hover:bg-indigo-300 disabled:opacity-60 disabled:bg-transparent"
        disabled={true}
        id="prev"
        name="prev"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
          />
        </svg>
      </button>
    </li>
    <li>
      <button
        className="active w-10 h-10 bg-white dark:bg-dark text-indigo-600 transition-colors duration-100 rounded-full focus:shadow-outline hover:bg-indigo-100 dark:hover:bg-indigo-300"
        id={1}
        onClick={[Function]}
      >
        1
      </button>
    </li>
    <li>
      <button
        className="flex items-center justify-center w-10 h-10 text-indigo-600 transition-colors duration-150 rounded-full focus:shadow-outline hover:bg-indigo-100 dark:hover:bg-indigo-300 disabled:opacity-60 disabled:bg-transparent"
        disabled={true}
        id="next"
        name="next"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M294.1 256L167 129c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.3 34 0L345 239c9.1 9.1 9.3 23.7.7 33.1L201.1 417c-4.7 4.7-10.9 7-17 7s-12.3-2.3-17-7c-9.4-9.4-9.4-24.6 0-33.9l127-127.1z"
          />
        </svg>
      </button>
    </li>
    <li>
      <button
        className="flex items-center justify-center w-10 h-10 text-indigo-600 transition-colors duration-150 rounded-full focus:shadow-outline hover:bg-indigo-100 dark:hover:bg-indigo-300 disabled:opacity-60 disabled:bg-transparent"
        disabled={true}
        id="last"
        name="last"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M10.293 15.707a1 1 0 010-1.414L14.586 10l-4.293-4.293a1 1 0 111.414-1.414l5 5a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0z"
            fillRule="evenodd"
          />
          <path
            clipRule="evenodd"
            d="M4.293 15.707a1 1 0 010-1.414L8.586 10 4.293 5.707a1 1 0 011.414-1.414l5 5a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0z"
            fillRule="evenodd"
          />
        </svg>
      </button>
    </li>
  </ul>
</nav>
`;
